istioctl/cmd/wait.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/api/meta"
+
+	"istio.io/istio/istioctl/pkg/kubernetes"
+	v2 "istio.io/istio/pilot/pkg/proxy/envoy/v2"
+)
+
+var (
+	forDistribution bool
+	targetResource  string
+	forDelete       bool
+	threshold       float32
+	timeout         time.Duration
+	resourceVersion string
+	//resourceVersions []string
+	rvLock sync.Mutex
+)
+
+const pollInterval = time.Second
+
+// waitCmd represents the wait command
+func waitCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "wait",
+		Short: "Wait for an Istio Resource",
+		Long: `A longer description that spans multiple lines and likely contains examples
+and usage of using your command. For example:
+
+Cobra is a CLI library for Go that empowers applications.
+This application is a tool to generate the needed files
+to quickly create a Cobra application.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			var versionChan chan string
+			var g *errgroup.Group
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			if resourceVersion == "" {
+				versionChan, g = getAndWatchResource(ctx, targetResource) // setup version getter from kubernetes
+			} else {
+				versionChan = make(chan string, 1)
+				versionChan <- resourceVersion
+			}
+			// wait for all deployed versions to be contained in resourceVersions
+			t := time.NewTicker(pollInterval)
+			resourceVersions := make([]string, 1)
+			for {
+				select {
+				case newVersion := <-versionChan:
+					resourceVersions = append(resourceVersions, newVersion)
+				case <-t.C:
+					finished, err := poll(resourceVersions)
+					if err != nil {
+						return err
+					} else if finished {
+						return nil
+					}
+				case <-ctx.Done():
+					// I think this means the timeout has happened:
+					t.Stop()
+					if err := g.Wait(); err != nil {
+						return err
+					}
+					return errors.New("timeout expired before resource was distributed")
```

Comment:
How about "timed out before resource %q became effective on all sidecars".  Change this for all the arguments ... show the user what we think the input was, let them know we are waiting for it to become effective on sidecars rather than merely "distributed".